File: static/js/rut.js
document.addEventListener("DOMContentLoaded", () => {
  document.querySelectorAll("input[data-rut]").forEach((input) => {
    const feedback = document.createElement("small");
    feedback.className = "field-feedback";
    feedback.setAttribute("aria-live", "polite");
    input.insertAdjacentElement("afterend", feedback);

    input.addEventListener("input", () => {
      const compact = input.value.replace(/[^0-9kK]/g, "").toUpperCase();
      if (compact.length < 2) {
        feedback.textContent = "";
        return;
      }
      const number = compact.slice(0, -1);
      const verifier = compact.slice(-1);
      const total = [...number].reverse().reduce((sum, digit, index) => sum + Number(digit) * (index % 6 + 2), 0);
      const expected = "0123456789K"[(11 - total % 11) % 11];
      feedback.textContent = number && verifier === expected ? "RUT válido" : "El dígito verificador no coincide.";
      feedback.classList.toggle("invalid", number !== "" && verifier !== expected);
    });

    input.addEventListener("blur", () => {
      const compact = input.value.replace(/[^0-9kK]/g, "").toUpperCase();
      if (compact.length < 2) return;
      const number = compact.slice(0, -1).replace(/^0+/, "") || "0";
      input.value = Number(number).toLocaleString("es-CL").replace(/,/g, ".") + "-" + compact.slice(-1);
    });
  });
});

File: organizations/rut.py
import re

from django.core.exceptions import ValidationError


def normalize_rut(value):
    compact_rut = re.sub(r"[^0-9kK]", "", value or "").upper()
    if len(compact_rut) < 2 or not compact_rut[:-1].isdigit():
        raise ValidationError("Ingrese un RUT chileno válido.")
    number, verifier = compact_rut[:-1].lstrip("0") or "0", compact_rut[-1]
    total = sum(int(digit) * factor for digit, factor in zip(reversed(number), [2, 3, 4, 5, 6, 7] * 2))
    if verifier != "0123456789K"[(-total) % 11]:
        raise ValidationError("El dígito verificador del RUT no coincide.")
    return f"{int(number):,}".replace(",", ".") + f"-{verifier}"
